kernel/interrupt.rs

```diff
@@ -1,48 +1,52 @@
 //! Interrupt handling.
 
-use alloc::boxed::Box;
-use core::mem::MaybeUninit;
+use alloc::{boxed::Box, vec::Vec};
 use kerla_runtime::{arch::enable_irq, spinlock::SpinLock};
-use kerla_utils::bitmap::BitMap;
 
 use crate::{deferred_job::run_deferred_jobs, interval_work};
 
-fn empty_irq_handler() {}
-
 type IrqHandler = dyn FnMut() + Send + Sync;
-const UNINITIALIZED_IRQ_HANDLER: MaybeUninit<Box<IrqHandler>> = MaybeUninit::uninit();
-static IRQ_HANDLERS: SpinLock<[MaybeUninit<Box<IrqHandler>>; 256]> =
-    SpinLock::new([UNINITIALIZED_IRQ_HANDLER; 256]);
-static ATTACHED_IRQS: SpinLock<BitMap<32 /* = 256 / 8 */>> = SpinLock::new(BitMap::zeroed());
+const NUM_IRQ_NUMBERS: usize = 256;
 
-pub fn init() {
-    let mut handlers = IRQ_HANDLERS.lock();
-    for handler in handlers.iter_mut() {
-        handler.write(Box::new(empty_irq_handler));
-    }
+/// Holds the interrupt handlers. The index is the IRQ number and this vector
+/// contains `NUM_IRQ_NUMBERS` entries.
+static IRQ_VECTORS: SpinLock<Vec<IrqVector>> = SpinLock::new(Vec::new());
+
+pub struct IrqVector {
+    handlers: Vec<Box<IrqHandler>>,
```

Comment:
This is going to be vector of vectors and is going to hurt the cache badly. I'd suggest, leaving array:
```Rust
static IRQ_VECTORS: SpinLock<[Vec<IrqVector>; 256]> = SpinLock::new([Vec::new(); 256]);
```
so at least you'd have static array in fixed memory location without relocations on new base interrupt vector.

Reply:
It sounds a sort of premature optimization. I do know it's not good for performance, but I don't want to consider too much time for it.

Reply:
I wouldn't call care for cache coherency in OS a premature optimisation, while Vector of Vector is well known for being bad at it. But, it's your call, @nuta.

Reply:
Moreover, original Linux Kernel uses a list for shared interrupts. The nice summary is here:
https://unix.stackexchange.com/questions/47306/how-does-the-linux-kernel-handle-shared-irqs

Then I wonder if it would be better to use IDT on x86_64 to put some handlers directly there, but that might be hard in Rust without some assembler trickery.

Reply:
I think we should not care until we actually need it. It's fast enough for me and I'd focus on adding more new feature at this early stage.

That said, I don't mean to give up the performance, in the safe way. Let's gradually consider the good approach to improve the performance by writing carefully written data structures in Rust 😃